### ISSUE TITLE
Stronger types for actions

### ISSUE THREAD

**Issue by zats** [ReSwift/ReSwift#8]
Hi, I wanted to start conversation about introducing stronger types into the framework.
I implemented a test app using the latest version of the framework to get a feel of it. Also I started playing with the barebones version implementation of the framework in the playground ([Flow.playground.zip](https://github.com/Swift-Flow/Swift-Flow/files/70506/Flow.playground.zip)).
My core objective for now is to make action type stronger. Also I experimented with implementing it as an enum, and, while can't be enforced on the level of type system, it exposes several advantages comparing to the string type + dictionary payload approach:

0. Autocompletion, makes typos impossible
1. Looking at action object, it is quite easy to guess what are the possible actions
2. It's also quite easy to know if you handled all possible action*
3. Most important one, associated values, allows to strongly type payload.

Also it has several disadvantages: 

1. Enum with many cases. In any sufficiently large app it might become a huge file (since currently it's not possible to extend existent enum)
2. Same problem from a different perspective: reducers are unlikely to handle all the possible actions in one function (which would be impractical). It means in practice we will end up using `default:` case or `if case let … = state` to handle a particular case.

Here is a gist of my simplified implementation:
* Action is a generic type, experiment with implementing it using enum with associated values for payload
* Concrete store is implemented as a generic class over state and action types
* Use `typealias` functions where possible (reducers) instead of protocols 
* I was trying to avoid `import Foundation` but there is still need for libdispatch (however it seems to be ported to linux)

My implementation misses many important features that you have in yours. 
But I wanted to get the conversation going with this little sample:)

**Comment by Ben-G** [comment]
Thanks a lot for your input! I actually started out using actions that were implemented as enums with associated types, I also experimented briefly with having a store that is generic over the action type.

The main reason I moved away from this model was to support serialization of actions. That enables hot reloading and time traveling even between app launches and terminations.

The problem with limited reflection in Swift is that I couldn't find a way to have typed actions **and** instantiate them through deserialization. One thing I have considered, is switching to `NSObject` subclasses for actions until Swift's serialization capabilities are improved.

Another alternative would be making the serialization of actions optional, since not all applications will need to be able to persist actions across app launches.

I would be curious on your thoughts regarding strongly typed and serializable actions.

**Comment by zats** [comment]
Great point! Give me couple of days :smiley:

**Comment by zats** [comment]
Hi Benjamin, I extended my playground with a recorder.

Basically, with enums, it is a manual process of making sure you serials both cases and associated values. But, somehow it still feels okay.  
Saying that, the approach I took does not prohibit you from going a [Mantle](https://github.com/Mantle/Mantle)-like way: making you action an `NSObject` object subclass, leveraging Objective-C runtime to infer and serialise all the properties.

As for the recorder, I implemented it as a two part system to separate responsibilities:

1. Recording reducer, when added to the list of reducers will persist all the actions
2. Extension function on the store will replay all actions found in a directory as long as the Action conforms to `Serializable` protocol. 

Some technical details:
Replaying stored actions is implemented in the least intrusive way possible, it means that the store is not aware of the fact that it replays "old" actions and it will treat them as if user dispatched them. Therefore it will record them again if you have recording reducer included into your reducers list. 
I don't see it as a critical problems since I thought of two possible workarounds (and I'm sure there are even more elegant ones):

1. Allow modifying list of reducers after Store initialisation (I'm thinking it might be a good idea regardless). Start without recording reducer, add it once replay is done.
2. More hacky solution is to point recording reducer to a temporary folder while replaying actions, switch it to point back to the original folder once replay is done

Overall replay feature, while I'm sure is useful seem to expose a problem of ever-growing actions list, so, being a development oriented feature, you will probably end up purging the stored action files or trim it somehow.

**Comment by Ben-G** [comment]
Hey there! Thanks for getting back so quickly with more input. When I responded to your original issue I was on the go, so the response was briefer than it should have been. I'll try to address all comments from both of your comments here.

**Sharing Code for API Proposals**

> P.S. let me know what is your preferred way of checking out the code, I reckon zip files is not the most convenient way to go

[I have created a separate repo within this organization where we can work on API proposals and examples.](https://github.com/Swift-Flow/API-Playgrounds/blob/master/StrongerTypedActions/zats/Flow%203.playground/Contents.swift) I've invited you as a member so you should be able to simply push code there. Feel free to change the folder structure as needed.

**Typed Actions**

> My core objective for now is to make action type stronger. Also I experimented with implementing it as an enum, and, while can't be enforced on the level of type system, it exposes several advantages comparing to the string type + dictionary payload

I didn't address this in my first response. I'm actually currently not using actions with types and dictionaries, instead I'm using typed actions that can be converted into these plain actions. 
[I've briefly discussed the approach in the docs over here](https://github.com/Swift-Flow/Swift-Flow/blob/master/Readme/GettingStarted.md#using-typed-actions). However, I don't have an example of these typed actions within these repos. 

[This project](https://github.com/Ben-G/Meet) is planned to be a reference for a real-world app being built with Swift Flow - it's currently still in a messy state - but it does show how to use typed actions and some other advanced features.

**TL;DR;** with the current action system, actions are untyped when they arrive in reducers, where you switch over the `type` string and then "cast" them into typed actions by initializing them. 

[Here's an example from the location reducer in the example app.](https://github.com/Ben-G/Meet/blob/master/Meet/Meet/Reducers/LocationServiceReducer.swift#L14-L25)

> 
1. Autocompletion, makes typos impossible
2. Looking at action object, it is quite easy to guess what are the possible actions
3. It's also quite easy to know if you handled all possible action*
4. Most important one, associated values, allows to strongly type payload.

This means the current action system addresses 1. and 4 - somewhat - even though it does involve a manual step of initializing the typed action from the untyped one within the reducer.

So I'll try to discuss your proposal from the angle where it mostly differs from the current API...

**Generic Store with Specified Action Type which is Enum**

Your API Proposal defines the possible actions as part of the store definition:

```
enum Action {
    case Add(URL: String)
    case Remove(index: Int)
}

Store<Action, AppState>
```

You bring up a potential downside:

> Enum with many cases. In any sufficiently large app it might become a huge file (since currently it's not possible to extend existent enum)

I think this can indeed be an issue. In my current example app I like to divide the action definitions into different categories, grouped by the domain they affect (Authentication, specific Scenes, etc.). Having all in a single enum could make it hard to organize large amount of actions.

One of the biggest issues of having the `Store` being generic over the action type, is that no component outside of the app dispatch actions to the `Store`.

Currently, e.g. [Swift Flow Router](https://github.com/Swift-Flow/Swift-Flow-Router) is implemented by hooking into the application state, dispatching its own actions and maintaining its own state within the host application's state. It's possible that this isn't the right way to implement the router, but I feel like it will be useful for some extensions to be able to dispatch actions to the store of the host application. If all the possible actions need to be defined within the host application that's no longer possible.

So I think my two main concerns are:

1. Developers would be forced to maintain one large enum, they wouldn't have the choice to use structs as action if the store was generic over one specific action type
2. Extensions wouldn't be able to dispatch actions (not :100: if this is an actual concern or a good thing yet at this point)

I feel like especially 1. would place a pretty heavy restriction on how this library can be used. 

**Other ways to improve the current ways action are handled?**

I'm not entirely sure about your feelings of how typed actions are currently used, but my two main concerns are:

1. In reducers developers are required to manually initialize a typed action from an untyped one
2. Developers need to provide action serialization/deserialization code

For me personally the 2. is a larger concern than the 1. That's because, even though it seems risky to try to create a typed action from an untyped one, practically this operation should never fail unexpectedly.

The only reason for unexpected failure would be that an action was serialized with an outdated version of properties and we're now trying to create the current action from this outdated serialized one. Since recording and replaying will likely only happen at development time this becomes even less of an issue.

*So IMO it seems most important to tackle this:*
> 2. Developers need to provide action serialization/deserialization code

*For aforementioned reasons I don't think that a single action enum is the best solution, but I'm very open to different opinions and a discussion around this!*

**Time Travel**

I like some of the points you bring up about record/replay a lot:

> Replaying stored actions is implemented in the least intrusive way possible, it means that the store is not aware of the fact that it replays "old" actions and it will treat them as if user dispatched them. 

Yes, the recorder should definitely become less intrusive, my current implementation is *very* hacky. I think you're headed in the right direction, however we should likely implement the recording of actions in a *middleware* (API being developed here: #9). In the redux philosophy, reducers should always be entirely side-effect free. A middleware provides you a way of extending the `dispatch` behavior, e.g. recording all dispatched actions. That way you can implement your idea without inducing side effects into reducers.

This technique allows you to address this problem:

> [...] Therefore it will record them again if you have recording reducer included into your reducers list.

When replaying actions you can lift a plain `Action` into a `RecordedAction` type - the middleware can then implement different behavior for the `Action` and `RecordedAction` type, recording the one and simply dispatching the other.

This is somewhat along the lines of how the original [Redux-Devtools](https://github.com/gaearon/redux-devtools) implementation works. @zats I would love to see a similar sophisticated approach for [Swift Flow Recorder](https://github.com/Swift-Flow/Swift-Flow-Recorder) (which we should likely rename to Swift Flow Dev Tools) if you have any time & interest on getting that started!

**Summary**

I tried hard to write less...

IMO we should tackle the following two issues discussed here:

1. **Make Action Serialization and Deserialization more convenient for developers**
2. **Implement tooling from time traveling similar to Redux-Devtools**

P.S. [Here's] (https://github.com/rackt/redux/issues/992) some very interesting reading regarding typed/untyped actions in the original Redux implementation...

**Comment by Ben-G** [comment]
@zats I will close this for now, since I believe most issues should be addressed with #12. Happy to reopen if we want to continue the discussion!

**Comment by zats** [comment]
Yes, I'm sorry for abandoning it :flushed: 
Congrats on joining the efforts with Karl though! :tada:
